arch/arm/core/aarch32/cpu_idle.S

```diff
@@ -62,19 +62,36 @@ SECTION_FUNC(TEXT, arch_cpu_idle)
 #endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
 #endif /* CONFIG_TRACING */
 
-#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE) \
-	|| defined(CONFIG_ARMV7_R)
-	cpsie i
-#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
-	/* clear BASEPRI so wfi is awakened by incoming interrupts */
+#if defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
```

Comment:
@stephanosio for the non-Cortex-M Mainline case, I do not feel it is OK to rely on the called to disable interrupts; we should do our own CPSID here, to be independent.

It is unlikely that irq_lock is going to change, but I don't want a dependence on what arch_irq_lock is doing.

Reply:
One of the reasons for keeping interrupts locked until the processor wakes up, in case of invocation from the idle thread, is to ensure that `_isr_wrapper` is not executed before going to sleep.

This is necessary because the idle thread sets `_kernel.idle`, and the `_isr_wrapper` checks this value and calls `z_sys_power_save_idle_exit` if set to a non-zero value. Locking interrupts before setting `_kernel.idle` and unlocking after waking up ensures that `z_sys_power_save_idle_exit` is executed only after the processor wakes up.

When `k_cpu_idle` (`arch_cpu_idle`) is called from a context other than the idle thread, as long as the caller does not set `_kernel.idle`, there should not be any problem. If the caller sets `_kernel.idle`, it would be the caller's responsibility to lock interrupts before doing so (adding `CPSID` here does not help in this case).

Reply:
Ah, right I see the point.

The problem is: right now there is an inconsistency between Baseline and Mainline.
What you describe here:

>When k_cpu_idle (arch_cpu_idle) is called from a context other than the idle thread, as long as the caller does not set _kernel.idle, there should not be any problem. If the caller sets _kernel.idle, it would be the caller's responsibility to lock interrupts before doing so (adding CPSID here does not help in this case).

...also holds for Mainline, but here you do call CPSID.

It is hard for the reader to understand why CPSID is not required for Baseline. We need to either add it, or then add a comment about why it is only added for v7-M

Reply:
> ...also holds for Mainline, but here you do call CPSID.

The CPSID for Mainline is there because Zephyr uses BASEPRI for interrupt (un)locking, and this value dictates the wake-up interrupt priority.

Since we must set BASEPRI to the lowest priority level to ensure proper wake-up, and this will certainly cause interrupts to be serviced unless we set `PRIMASK`, we must `CPSID`.

In summary, this is only an artifact of the current Zephyr Cortex-M Mainline interrupt management implementation, nothing more.

Reply:
This seems related to my notes when trying to understand zephyr system power management recently:

> `k_cpu_idle()` ... must unlock the `arch_irq_lock()` taken by the idle thread when it's active, even though that API is documented as callable only when there is no lock.  Apparently that really means no spinlock.

Or maybe no `irq_lock`.  Since the difference between them is rather subtle (#12537).

I haven't looked closely at this PR yet to determine if that confusion remains, but if it does it'd be nice to eliminate it.

Reply:
>The CPSID for Mainline is there because Zephyr uses BASEPRI for interrupt (un)locking, and this value dictates the wake-up interrupt priority.
Since we must set BASEPRI to the lowest priority level to ensure proper wake-up, and this will certainly cause interrupts to be serviced unless we set PRIMASK, we must CPSID.
In summary, this is only an artifact of the current Zephyr Cortex-M Mainline interrupt management implementation, nothing more.

Sure, I only want this to be documented for ease of reading ( as I suggested above, but maybe not to clear :) )

Reply:
> Sure, I only want this to be documented for ease of reading ( as I suggested above, but maybe not to clear :) )

@ioannisg I will try to think of a nice way to describe this without going into too much details, as that would probably confuse the reader even more.

Reply:
Added explanation for non-BASEPRI case.

Reply:
thanks